Add QSim device

thisac opened the pull request:
Adds a new [QSim device](https://github.com/quantumlib/qsim/blob/master/docs/cirq_interface.md) by importing `qsimcirq` and creating a QSim circuit using a Cirq circuit as `qsimcirq.QSimCircuit(cirq_circuit=cirq.Circuit())` and then using `qsimcirq.QSimSimulator`. 

It can be installed with `pip install qsimcirq`.

thisac commented:
Quick benchmark for the following circuit:

```python
def circuit(x, y, z):
    for j in range(i-1):
        qml.RZ(z, wires=[j])
        qml.RY(y, wires=[j])
        qml.RX(x, wires=[j])
        qml.CNOT(wires=[j, j+1])

    return qml.expval(qml.PauliZ(wires=i-1))
```

![image](https://user-images.githubusercontent.com/6934626/89346039-6f4ca380-d676-11ea-9bd9-9773964e46d4.png)

co9olguy commented:
Nice @thisac, good to see this making progress!

josh146 commented:
Awesome @thisac! 🎉 

> Note: this branch currently imports qsimcirq and thus assumes that it is installed. We probably either want to have some sort of check whether the user has qsimcirq installed or add it to the requirements.

I had a look at the qsim PyPI package, and it looks like it includes platform-specific code (C++), but binary wheels are not provided. Since it could be hard for users to install, I suggest not adding it to `setup.py`, but taking the same approach we do with optional dependencies in PL --- use a `try`/`except` around the import, and make sure the qsim import doesn't occur on other device loads!

trbromley commented on pennylane_cirq/simulator_device.py:
```diff
@@ -132,7 +132,8 @@ def apply(self, operations, **kwargs):
         # We apply an identity gate to all wires, otherwise Cirq would ignore
         # wires that are not acted upon
-        self.circuit.append(cirq.IdentityGate(len(self.qubits))(*self.qubits))
+        for qb in self.qubits:
+            self.circuit.append(cirq.IdentityGate(1)(qb))
```
So in the new approach we're looping over individual qubits, but before we passed all of the qubits to a single identity gate?

Was this done for clarity or was it broken?

thisac replied:
@trbromley It didn't work applying a single identity gate, so this was the simplest work-around I could think of. I noticed that it's actually done for the `SimulatorDevice` as well during the wires refactor, but I'm not sure about that change exactly.

trbromley commented on pennylane_cirq/qsim_device.py:
```diff
@@ -0,0 +1,111 @@
+import cirq
+import numpy as np
+
+try:
+    import qsimcirq
+except ImportError:
+    raise ImportError(
+        "QSim Cirq is needed for the QSim devices to work."
+        "\nIt can be installed using pip:"
+        "\n\npip install qsimcirq"
+    )
```
Are we avoiding making this a dependency as `qsimcirq` might be difficult to install?

thisac replied:
Yes, I think that was the case. I didn't have any issues myself, but I believe @josh146 mentioned something about this being potentially difficult to build.

trbromley commented on pennylane_cirq/qsim_device.py:
```diff
@@ -0,0 +1,111 @@
+
+class QSimDevice(SimulatorDevice):
+    r"""QSim device for PennyLane.
+    Args:
```
```suggestion

    Args:
```

trbromley commented on pennylane_cirq/qsim_device.py:
```diff
@@ -0,0 +1,111 @@
+        self.circuit = qsimcirq.QSimCircuit(cirq_circuit=cirq.Circuit())
+        self._simulator = qsimcirq.QSimSimulator()
+
+        self._initial_state = None
+        self._result = None
+        self._state = None
```
Do we need these, since they are initialized in `QSimDevice.super()`?

thisac replied:
Oh, you're right. Removed them now. Thanks @trbromley! :tada:

trbromley commented on pennylane_cirq/qsim_device.py:
```diff
@@ -0,0 +1,111 @@
+
+class QSimhDevice(SimulatorDevice):
+    r"""QSimh device for PennyLane.
+    Args:
```
```suggestion

    Args:
```

trbromley commented on pennylane_cirq/qsim_device.py:
```diff
@@ -0,0 +1,111 @@
+        self.circuit = cirq.Circuit()
+        self._simulator = qsimcirq.QSimhSimulator(qsimh_options)
+
+        self._initial_state = None
+        self._result = None
+        self._state = None
```
Similarly, maybe we don't need these

trbromley commented on pennylane_cirq/qsim_device.py:
```diff
@@ -0,0 +1,111 @@
+        qubits (List[cirq.Qubit]): A list of Cirq qubits that are used
+            as wires. The wire number corresponds to the index in the list.
+            By default, an array of ``cirq.LineQubit`` instances is created.
+        qsimh_options (dict): A dictionary with options for the QSimh simulator. See the `QSim
+            usage documentation <https://github.com/quantumlib/qsim/blob/master/docs/usage.md>`__
+            for further details.
```
This is probably the plan already, but would be good to give a nice explanation of this in the docs.

trbromley commented on pennylane_cirq/qsim_device.py:
```diff
@@ -0,0 +1,111 @@
+        super().__init__(wires, shots, analytic, qubits)
+        if qsimh_options is None:
+            qsimh_options = {
+                'k': [0],
+                'w': 0,
+                'p': 0,
+                'r': 1
```
Are these options required or could we use a qsimh default?

trbromley replied:
Also, would it make sense to make `qsimh_options` a required argument in this device? I'm not sure why those numbers above are good, so feel it might be good to force the user to decide.

thisac replied:
I'm quite sure that these options are required. I just put in the default recommended ones from the website here.

I guess having the options as a required argument makes a lot of sense (since they are required by qsimh). :thinking:

trbromley replied:
Yes, I would vote to make `qsimh_options` a required arg, unless we're quite confident that those options make sense as a default.

trbromley commented on pennylane_cirq/qsim_device.py:
```diff
@@ -0,0 +1,111 @@
+                bitstrings=list(range(2 ** len(self.wires)))
+            )
+
+            self._state = np.array(state)
```
Things piece together nicely!

trbromley commented on setup.py:
```diff
@@ -35,7 +35,10 @@
+                                           "cirq.mixedsimulator = pennylane_cirq:MixedStateSimulatorDevice",
+                                           "cirq.qsim = pennylane_cirq.qsim_device:QSimDevice",
+                                           "cirq.qsimh = pennylane_cirq.qsim_device:QSimhDevice",
+                                           "cirq.qflex = pennylane_cirq.qflex_device:QFlexDevice"],},
```
```suggestion
```

trbromley reviewed (commented):
Thanks @thisac, looks great so far and also works on a simple prototype I made! I'll be happy to take another look when the tests and supporting docs are added.
:100:

thisac commented on tests/test_qsim_device.py:
```diff
@@ -0,0 +1,382 @@
+# Copyright 2018 Xanadu Quantum Technologies Inc.
```
The tests in this file have been copied from `test_cirq_device.py`, with only some minor modifications to work with `QSimDevice`.

thisac commented on tests/test_qsimh_device.py:
```diff
@@ -0,0 +1,400 @@
+# Copyright 2018 Xanadu Quantum Technologies Inc.
```
Same with these tests. They have been copied from `test_cirq_device.py`, with only some minor modifications to work with `QSimhDevice`. In addition, two tests for the `qsimh_options` kwarg has been added.

thisac commented:
Sorry for taking so long to finish this. I've done some minor changes and added tests (mainly copied from the CirqDevice tests, with some minor alterations). Two things to note is that

1. qsimcirq doesn't support `ControlledGate` (i.e. gates constructed using the `controlled_by()` method), so those tests have simply been removed from the test-files for now. I might look into it tomorrow and see if any more tests can be added to replace these.

2. There seems to be an issue with `cirq.inverse(gate)` for gates `cirq.X`, `cirq.Y` and `cirq.Z` where it sets the exponent to an integer (-1), while qsim only assumes/checks for float exponents. Thus `PauliX.inv()`, `PauliY.inv()`, `PauliZ.inv()` raise fairly obscure C++ errors when used with qsim and qsimh.

josh146 commented on pennylane_cirq/qsim_device.py:
```diff
@@ -0,0 +1,110 @@
+    def __init__(self, wires, shots=1000, analytic=True, qubits=None, qsimh_options=None):
+        super().__init__(wires, shots, analytic, qubits)
+        if qsimh_options is None:
+            qsimh_options = {"k": [0], "w": 0, "p": 0, "r": 1}
```
The qsimh options are quite opaque 😆 You don't often see so many single letter dictionary keys

josh146 commented on pennylane_cirq/qsim_device.py:
```diff
@@ -0,0 +1,110 @@
+    def apply(self, operations, **kwargs):
+        CirqDevice.apply(self, operations, **kwargs)
+
+        # TODO: remove the need for this hack by keeping better track of unused wires
+        # We apply identity gates to all wires, otherwise Cirq would ignore
+        # wires that are not acted upon
```
Is this todo copied from the existing Cirq device? Looks familiar

thisac replied:
Yes, it was written differently before, but I noticed the `TODO:` in the existing Cirq device, so I thought it might be good to duplicate it here so we don't forget about this one as well (unfortunately this  part needed to be duplicated).

thisac replied:
@josh146 Would you recommend just keeping the `TODO`s here and simply ignore the pylint complaint for this PR?

josh146 replied:
Yep, fine to simply ignore pylint here!

josh146 commented on pennylane_cirq/simulator_device.py:
```diff
@@ -135,6 +135,7 @@ def apply(self, operations, **kwargs):
         # TODO: remove the need for this hack by keeping better track of unused wires
```
I suppose this answers my question above!

josh146 commented on setup.py:
```diff
@@ -35,7 +35,9 @@
     "entry_points": {"pennylane.plugins": ["cirq.simulator = pennylane_cirq:SimulatorDevice",
-                                           "cirq.mixedsimulator = pennylane_cirq:MixedStateSimulatorDevice"],},
+                                           "cirq.mixedsimulator = pennylane_cirq:MixedStateSimulatorDevice",
+                                           "cirq.qsim = pennylane_cirq.qsim_device:QSimDevice",
+                                           "cirq.qsimh = pennylane_cirq.qsim_device:QSimhDevice"],},
```
😍

josh146 reviewed (commented):
Looks good @thisac! Will be awesome to have this merged in 💪 

The final part of this PR would be to add documentation for the two new devices (i.e., by making `doc/devices/qsim.rst` and `qsimh.rst`, adding cards to `index.rst`, and adding the new pages to the TOC).

This would be the place to discuss:

* The reasons to pick qsim/qsimh over the Cirq simulator (could even include the benchmarking plot here!)

* Any additional installation instructions

* Minimal code examples

* Details on any device options (I think relevant for qsimh --- doesn't have to be too detailed, and can link back to the qsim docs)

* Any 'gotchas' the user should be aware of

trbromley commented on pennylane_cirq/qsim_device.py:
```diff
@@ -0,0 +1,110 @@
+
+    def __init__(self, wires, shots=1000, analytic=True, qubits=None):
+        super().__init__(wires, shots, analytic, qubits)
+        self.circuit = None
```
For my understanding, why not `self.circuit = qsimcirq.QSimCircuit(cirq_circuit=self.circuit)` here? I guess it makes more sense in `reset()`?

thisac replied:
It was there before, but following the logic in the Cirq device, I changed it to be initialized as `None`. I noticed it since there were tests testing for specifically that. Not sure if there's a reason for having it this way, but as long as `reset` is used the way it is now, it should work equally well.

trbromley replied:
Thanks!

trbromley commented on pennylane_cirq/qsim_device.py:
```diff
@@ -0,0 +1,110 @@
+        self._simulator = qsimcirq.QSimhSimulator(qsimh_options)
+
+    def apply(self, operations, **kwargs):
+        CirqDevice.apply(self, operations, **kwargs)
```
Just for my understanding: there is something in the `SimulatorDevice` version of `apply()` that we don't want to use, preventing us from calling `super().apply()` here?

Would it make sense for this device to inherit directly from `CirqDevice`? (maybe not, perhaps `SimulatorDevice` is useful in other places).

thisac replied:
Yes, that's exactly the case. `SimulatorDevice.apply` uses the method `_simulator.simulate` which qsimh doesn't have (qsim has it, and thus simply uses `SimulatorDevice.apply`). 

> Would it make sense for this device to inherit directly from CirqDevice? (maybe not, perhaps SimulatorDevice is useful in other places).

Not really, since quite a lot of the methods in `SimulatorDevice` are needed (e.g. `Cirq.apply` calls several abstract methods that only exist in `SimulatorDevice`). The alternative would be to make qsimh inherit from `CirqDevice` and rewrite many of the functions from `SimulatorDevice`, but I'm not sure if I think that actually makes things better. It might seem more structured, but would mean maintaining two version of many of the functions. I was thinking about this while writing it, and decided on the current solution. :slightly_smiling_face:

trbromley replied:
Perfect, thanks for the explanation!

trbromley commented on tests/test_qsim_device.py:
```diff
@@ -0,0 +1,382 @@
+from pennylane_cirq.qsim_device import QSimDevice
+
+
+@patch.multiple(QSimDevice, __abstractmethods__=set())
```
Never seen that before!!

trbromley replied:
What does it do :laughing:

josh146 replied:
This is a way of allowing you to instantiate an abstract class for testing 🙂 Essentially, the `__abstractmethods__` method is temporarily patched to an empty set, so that the class looks to have no abstract methods

trbromley commented on tests/test_qsim_device.py:
```diff
@@ -0,0 +1,382 @@
+        assert len(dev.qubits) == 4
+        assert dev.qubits == qubits
+
+    def test_outer_init_of_qubits_unordered(self):
+        """Tests that giving qubits as parameters to QSimDevice works when the qubits are not ordered consistently with Cirq's convention."""
+
+        qubits = [
+            cirq.GridQubit(0, 1),
+            cirq.GridQubit(1, 0),
+            cirq.GridQubit(0, 0),
+            cirq.GridQubit(1, 1),
+        ]
+
+        dev = QSimDevice(4, 100, False, qubits=qubits)
+        assert len(dev.qubits) == 4
+        assert dev.qubits == sorted(qubits)
```
Nice!

trbromley commented on tests/test_qsim_device.py:
```diff
@@ -0,0 +1,382 @@
+            dev = QSimDevice(3, 100, False, qubits=qubits)
+
+
+class TestQSimDeviceIntegration:
```
Are these integration tests?

trbromley commented on tests/test_qsim_device.py:
```diff
@@ -0,0 +1,382 @@
+
+        qsim_device_1_wire.reset()
+
+        # try:
+        qsim_device_1_wire.apply([gate])
+        # except:
+            # sdfsd
```
```suggestion
        qsim_device_1_wire.apply([gate])
```

trbromley commented on tests/test_qsimh_device.py:
```diff
@@ -0,0 +1,400 @@
+from pennylane_cirq.qsim_device import QSimhDevice
+
+
+@patch.multiple(QSimhDevice, __abstractmethods__=set())
```
Unsure if we need this also here, isn't it just to help with `CirqDevice` which has some abstract methods. Do we still have abstract methods in `QSimhDevice`?

thisac replied:
Oh, my bad, I copied the tests for `CirqDevice` rather than for `SimulatorDevice`. Thanks for catching this @trbromley! :facepalm:

trbromley commented on tests/test_qsimh_device.py:
```diff
@@ -0,0 +1,400 @@
+        assert qsimh_options.items() <= dev._simulator.qsimh_options.items()
+
+
+class TestQSimhDeviceIntegration:
```
Also here, they still seem more like unit tests than integration :thinking:

trbromley commented on tests/test_qsimh_device.py:
```diff
@@ -0,0 +1,400 @@
+
+        qsimh_device_1_wire.reset()
+
+        # try:
+        qsimh_device_1_wire.apply([gate])
+        # except:
+            # sdfsd
```
```suggestion
        qsimh_device_1_wire.apply([gate])
```

thisac replied:
Another :facepalm:. I find this to be an easy way to get the tests to stop and enter the terminal for debugging. There's probably a better way, but this generally works quite nicely. :laughing:

trbromley reviewed (commented):
Thanks @thisac! I noticed in the tests that the devices don't support inverses and some controlled gates - we should probably update the capabilities to reflect that (e.g., `"inverse_operations" = True`) and maybe update the `_operation_map` by taking out the relevant entries?

trbromley commented on pennylane_cirq/qsim_device.py:
```diff
@@ -53,13 +54,44 @@ class QSimDevice(SimulatorDevice):
+
+    def _apply_basis_state(self, basis_state_operation):
+        # pylint: disable=missing-function-docstring
+        if not self.analytic:
+            raise qml.DeviceError("The operation BasisState is only supported in analytic mode.")
```
I'm curious why it doesn't work in non-analytic mode?

thisac replied:
Not sure actually. I copied it over from the SimlatorDevice class. :thinking:

josh146 replied:
(Is this true for qsim? or is it a Cirq restriction?)

josh146 replied:
It might be worth removing `BasisState` from the supported operations, so that PL will always default to the decomposition (e.g., PauliX gates), which should always be supported.

thisac commented:
Thanks @trbromley  and @josh146 for the reviews. I realized that I'd copied over the wrong tests, so I've fixed that now. While doing that I noticed quite a few things that didn't work. I've added some temporary comments in the tests as well as updated the GitHub comments at the top of each test file, where I've detailed the issues.

I still have to update the docs; look through the tests again and add qsim/qsimh specific tests; make qsimh_options a required kwarg (I thinks that's a good idea, and simply clearing it up in the docs); and, finally, figure out what to do with the failing/commented out tests. Also, I'm not sure if/how qsim can let the integration tests know that it doesn't support `QubitStateVector` while still allowing it for the other Cirq devices. :thinking:

josh146 commented:
Thanks @thisac!

> Also, I'm not sure if/how qsim can let the integration tests know that it doesn't support QubitStateVector while still allowing it for the other Cirq devices. 🤔

Can you simply overwrite the `operations` attribute/property in the `QSimDevice` class, rather than relying on the inherited supported operations coming from the parent simulator device?

co9olguy commented:
I've removed my review request because I don't really have anything specific to offer on top of what @josh146  and @trbromley can already offer for code review.

I'd suggest if any questions/clarification comes up for how to deal with ordering, `LineQubit`s, `GridQubit`s, etc. in PennyLane-Cirq, then definitely reach out to me. I probably would understand best on those issues

thisac commented on doc/index.rst:
```diff
@@ -31,7 +31,17 @@ Currently, PennyLane-Cirq provides two Cirq devices for PennyLane:
     :link: devices/mixed_simulator.html
- 
+
+.. devicegalleryitem::
+    :name: 'cirq.qsim'
+    :description: The qsim simulator backend.
+    :link: devices/qsim.html
+
+.. devicegalleryitem::
+    :name: 'cirq.qsimh'
+    :description: The qsimh simulator backend.
+    :link: devices/qsim.html
```
Wasn't sure about having two cards link to the same page. I only added a single page (for both qsim and qsimh), since they work very similarly. Do you think it's better having two separate pages, or perhaps just keeping the qsim card here, even though technically qsimh is a separate device? :thinking:

thisac replied:
I think that might be the best way to go as well actually. I've updated it now (but kept the line `Currently, PennyLane-Cirq provides four Cirq devices for PennyLane:`, since there are still, technically, four devices).

josh146 commented on pennylane_cirq/qsim_device.py:
```diff
@@ -110,51 +104,38 @@ class QSimhDevice(SimulatorDevice):
-            self.circuit.append(cirq.X(q))
-
-    def _apply_qubit_state_vector(self, qubit_state_vector_operation):
+        ops = set(self._operation_map.keys()) - {
+            "QubitStateVector",
+            "BasisState",
+            "CRX",
+            "CRY",
+            "CRZ",
+            "CRot",
+        }
+        return ops
```
This is much cleaner!

josh146 commented on tests/test_qsim_device.py:
```diff
@@ -37,6 +37,51 @@ def test_device_loading(self):
+    @pytest.mark.parametrize("analytic", [True, False])
+    @pytest.mark.parametrize("shots", [8192])
+    def test_one_qubit_circuit(self, shots, analytic, tol):
+        """Test that devices provide correct result for a simple circuit"""
```
A good test to have 🙂

josh146 commented on tests/test_qsim_device.py:
```diff
@@ -173,14 +204,17 @@ def test_apply_operation_state_preparation(
         self, qsim_device_1_wire, tol, op, input, expected_output, par
     ):
-        """Tests that applying an operation yields the expected output state for single wire
+        """Tests that applying an operation yields the expected output probabilities for single wire
```
:+1:

josh146 reviewed (approved):
Nice work @thisac! I left a couple of suggestions in the documentation; once those are sorted it should be ready to go 🙂

thisac commented on doc/devices/qsim.rst:
```diff
@@ -0,0 +1,113 @@
+The qsim device
+===============
+
+qsim is a Schrödinger full state-vector simulator while qsimh is a hybrid
+Schrödinger-Feynman simulator, cutting the qubit lattice into two parts and then
+using qsim to simulate each part individually. For further details see the `qsim
+website <https://github.com/quantumlib/qsim>`__. These simulators are
+considerably faster compared to the Cirq simulator for many qubits, as can be
+seen in the benchmark below.
```
Thanks @josh146. This is much better. :100:

thisac commented on doc/devices/qsim.rst:
```diff
@@ -0,0 +1,113 @@
+^^^^^^^^^^^^^
+
+qsimh requires specific options to be set for the simulator. These can be passed
+by the positional argument ``qsimh_options``. See the `QSimh usage documentation
```
I've been super-confused whether to spell it QSim or qsim (since I've seen both in different places). They use qsim and qsimh on the website and in the supremacy paper, so I went with that. :laughing:

trbromley reviewed (approved):
Thanks @thisac, looks awesome! Left some suggestions on the docs.

thisac commented on doc/devices/qsim.rst:
```diff
@@ -0,0 +1,120 @@
+        qml.CNOT(wires=[0, 1])
+        return qml.expval(qml.PauliZ(wires=1))
+
+You can then execute the circuit like any other function to get the quantum mechanical expectation value.
```
```suggestion
You can then execute the circuit like any other function to get the quantum mechanical expectation value:
```

thisac commented on setup.py:
```diff
@@ -36,6 +36,8 @@
     "entry_points": {"pennylane.plugins": ["cirq.simulator = pennylane_cirq:SimulatorDevice",
                                            "cirq.mixedsimulator = pennylane_cirq:MixedStateSimulatorDevice",
+                                           "cirq.qsim = pennylane_cirq:QSimDevice",
+                                           "cirq.qsimh = pennylane_cirq:QSimhDevice"],},
```
```suggestion
                                           "cirq.qsimh = pennylane_cirq:QSimhDevice",

```

thisac commented on setup.py:
```diff
@@ -36,6 +36,9 @@
                                            "cirq.mixedsimulator = pennylane_cirq:MixedStateSimulatorDevice",
+                                           "cirq.qsim = pennylane_cirq:QSimDevice",
+                                           "cirq.qsimh = pennylane_cirq:QSimhDevice",
+
```
```suggestion
```

josh146 commented:
Thanks @thisac for sorting out the test issue. This looks ready to merge, but before we do so can I get you to update the changelog? You can use the same style as the Pasqal entry in the changelog (and use some of the intro on the `qsim.rst` usage page).

On another note: we should look into using [pytest-randomly](https://pypi.org/project/pytest-randomly/) to avoid this testing issue in future. In particular, pytest-randomly:

* Randomly shuffles the order of test items

* Resets random.seed() at the start of every test case and test to a fixed number
